### PR TITLE
Dynamic Consent from the DB instead local filesystem

### PR DESCRIPTION
**Pull request by raginirai553**

### CODE REVIEW

File: microsetta_private_api/db/patches/0104.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    account_id uuid NOT NULL,
+    PRIMARY KEY (consent_id)
+);
+CREATE TABLE ag.consent_audit (
+    signature_id uuid NOT NULL DEFAULT uuid_generate_v4(),
+    parent_1_name varchar(200),
+    parent_2_name varchar(200),
+    deceased_parent boolean,
+    assent_obtainer varchar(200),
+    consent_id uuid NOT NULL,
```
Comment (wasade) [inline microsetta_private_api/db/patches/0104.sql:17]: Should `consent_id` have a `UNIQUE` constraint?

File: microsetta_private_api/repo/consent_repo.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            else:
+                return _row_to_consent_document(r)
+
+    def is_consent_required(self, source_id, consent_type):
+        with self._transaction.dict_cursor() as cur:
+            cur.execute("SELECT " + ConsentRepo.signature_read_cols + " FROM "
+                        "ag.consent_audit INNER JOIN "
+                        "ag.consent_documents ON "
+                        "consent_audit.consent_id = "
+                        "consent_documents.consent_id "
```
Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:116]:
```suggestion
                        "USING(consent_id) "
```

File: microsetta_private_api/repo/consent_repo.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                        "WHERE consent_audit.source_id = %s and "
+                        "consent_documents.consent_type "
+                        "LIKE %s ORDER BY sign_date DESC",
```
Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:119]: What's the motivation for this to be glob like rather than an exact test?

File: microsetta_private_api/db/patches/0104.sql [hunk 4]
```diff
@@ -0,0 +1,22 @@
+CREATE TABLE ag.consent_documents (
+    consent_id uuid NOT NULL DEFAULT uuid_generate_v4(),
+    consent_type varchar(50) NOT NULL,
+    locale varchar(10) NOT NULL,
+    date_time TIMESTAMP WITH TIME ZONE NOT NULL,
+    consent_content varchar NOT NULL,
+    reconsent_required INT NOT NULL,
```
Comment (wasade) [inline microsetta_private_api/db/patches/0104.sql:7]: Should this be `boolean`?

File: microsetta_private_api/repo/consent_repo.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                        (source_id, ('%' + consent_type + '%')))
+
+            r = cur.fetchone()
+            if r is None:
+                return True
+            elif r['reconsent_required'] == 0:
```
Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:125]: If `reconsent_required` is made `boolean` in the database, then this could be changed to a direct test for true/false

Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:125]: I think it would be more clear to not use an `elif` here as the value being tested and its interpretation is quite different. 

```suggestion
            if r['reconsent_required'] == 0:
```

File: microsetta_private_api/db/patches/0104.sql [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    account_id uuid NOT NULL,
```
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_private_api/db/patches/0104.sql:8]: ...maybe I'm misunderstanding something here, but shouldn't the consent documents be independent of an account, and the `consent_audit` record who had signed which specific documents?

File: microsetta_private_api/repo/consent_repo.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            else:
+                consent_doc_type = r["consent_type"]
+                cur.execute("SELECT date_time FROM "
+                            "ag.consent_documents WHERE consent_type = %s "
+                            "ORDER BY date_time DESC LIMIT 1",
+                            (consent_doc_type,))
+
+                s = cur.fetchone()
+                if s is None:
+                    return True
```
Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:136]: I believe this implies that there is not a consent document which matches the requested `consent_doc_type`. Should this raise a `RepoException`?

File: microsetta_private_api/repo/consent_repo.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                else:
+                    sign_date = r["sign_date"]
+                    doc_date = s["date_time"]
+
+                    if doc_date > sign_date:
+                        return True
+                    else:
+                        return False
```
Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:144]:
```suggestion
                    return doc_date > sign_date
```

File: microsetta_private_api/repo/consent_repo.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                raise NotFound("Consent Document does not exist!")
+
+            sourceRepo = SourceRepo(self._transaction)
+            source_id = consent_signature.source_id
+
+            if sourceRepo.get_source(account_id, source_id) is None:
+                raise NotFound("Source does not exist!")
+
+            cur.execute("INSERT INTO consent_audit (" +
+                        ConsentRepo.signature_write_cols + ") "
```
Comment (wasade) [inline microsetta_private_api/repo/consent_repo.py:161]:
```suggestion
                        self.signature_write_cols + ") "
```

**Review by wasade — CHANGES_REQUESTED** [review]
This is looking pretty good. A few additional comments not scoped inline:

- please add unit tests for the repo logic, including evaluating edge cases
- please remove the `.DS_Store` files

@cassidysymons, please take a look at my comments particularly those with the database patch.

File: microsetta_private_api/api/tests/test_integration.py [hunk 10]
```diff
@@ -997,6 +1019,56 @@ def test_create_human_source(self):
...
+
+        with Transaction() as t:
+            consent_repo = ConsentRepo(t)
+            consent = ConsentDocument.from_dict(CONSENT_DOC,
+                                                DUMMY_ACCT.get("id"),
+                                                CONSENT_DOC_ID
+                                                )
+
+            consent_repo.create_doc(consent)
+            t.commit()
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/api/tests/test_integration.py:1043]: we needed this explicit data update as the admin side function is not yet built which would have take care of this, But I believe we should be able to delete the updates once the test passes, I will work on this.

File: microsetta_private_api/db/patches/0104.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (cassidysymons) [reply microsetta_private_api/db/patches/0104.sql:8]: The account_id is intended to store which admin user created a given consent document. The eventual goal is to have an admin interface to update account documents, taking it off of developers' plates. However, developing that was deferred until later since it's not necessary for re-launch.

However, two tweaks here:
1) It would make sense to provide a more descriptive name, such as admin_account_id.
2) We can't have the NOT NULL constraint since we will be adding consent documents via database batches until we can develop the admin interface.

File: microsetta_private_api/db/patches/0104.sql (hunk 1 above)
Comment (cassidysymons) [reply microsetta_private_api/db/patches/0104.sql:17]: Yes, there's no situation where a given source should be able to sign the same consent document twice.

File: microsetta_private_api/db/patches/0104.sql (hunk 4 above)
Comment (cassidysymons) [reply microsetta_private_api/db/patches/0104.sql:7]: +1 on boolean

File: microsetta_private_api/api/tests/test_integration.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (cassidysymons) [reply microsetta_private_api/api/tests/test_integration.py:1043]: We shouldn't commit any transactions in testing. Functions should be fully self-contained, so if a given test needs records in the database to exist, create them as the first steps in that test, then proceeding with the actual testing code.

File: microsetta_private_api/repo/consent_repo.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def _row_to_consent_document(r):
+    return ConsentDocument(
+        r["consent_id"],
+        r["consent_type"],
+        r["locale"],
+        r["date_time"],
+        r["consent_content"],
+        getattr(r, 'account_id', None),
```
[the remark this answers is not in the document]
Comment (cassidysymons) [reply microsetta_private_api/repo/consent_repo.py:26]: As I noted in the database batch comment, this is related to an admin account creating a new version of a consent doc, rather than a user signing the consent doc.

File: microsetta_private_api/db/patches/0104.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_private_api/db/patches/0104.sql:8]: What about having `NOT NULL` and associating them to my, Se Jin's or your account?

[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_private_api/db/patches/0104.sql:8]: It should still have a FK nonetheless

[the remark this answers is not in the document]
Comment (cassidysymons) [reply microsetta_private_api/db/patches/0104.sql:8]: Having one of our account IDs as a placeholder seems fine, although I don't love the idea of hardcoding that account ID into database patches. Can we set up a trigger to do it dynamically? Hardcoding one of our email addresses into a trigger seems less uncomfortable than hardcoding the account ID.

[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_private_api/db/patches/0104.sql:8]: I don't see `INSERT` logic here so why not just do it where those occur? Of the IDs, mine will be stable w/ the test database. A downside to a trigger is it's really special case. But, all that said, I don't have a strong opinion on the approach but I do feel strongly about non null if that makes sense :)

File: microsetta_private_api/api/_consent.py [hunk 12]
```diff
@@ -1,18 +1,68 @@
...
+        documents = consent_repo.get_consent_document(consent_id)
+        data = [x.to_api() for x in documents]
+
+    return jsonify(data), 200
+
+
+def check_consent_signature(account_id, source_id, consent_type, token_info):
+    _validate_account_access(token_info, account_id)
+
+    res = False
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/api/_consent.py:43]: Agreed, corrected.

File: microsetta_private_api/api/_consent.py [hunk 13]
```diff
@@ -1,18 +1,68 @@
...
+        consent_repo = ConsentRepo(t)
+        sign_id = str(uuid.uuid4())
+        consent_sign = ConsentSignature.from_dict(body, source_id, sign_id)
+        consent_repo.sign_consent(account_id, consent_sign)
+        t.commit()
+
+        response = jsonify({"result": True})
+        response.status_code = 201
+        response.headers['Location'] = '/api/accounts/%s' % \
+            (account_id)
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/api/_consent.py:66]: Corrected

File: microsetta_private_api/api/tests/test_api.py [hunk 14]
```diff
@@ -1379,6 +1384,68 @@ def test_source_update_success(self):
...
+
+        response = self.client.post(
+            '/api/accounts/%s/sources/%s/consent/%s' %
+            (dummy_acct_id, dummy_source_resp["source_id"], DATA_CONSENT),
+            content_type='application/json',
+            data=json.dumps(consent_data),
+            headers=self.dummy_auth)
+
+        self.assertEquals(201, response.status_code)
+
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/api/tests/test_api.py:1418]: Working on it.

File: microsetta_private_api/api/tests/test_integration.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/api/tests/test_integration.py:1043]: I have corrected the code to create sample consent document at the beginning of the test and delete the same at the end of the test.

File: microsetta_private_api/db/patches/0104.sql [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    PRIMARY KEY (consent_id)
+);
+CREATE TABLE ag.consent_audit (
+    signature_id uuid NOT NULL DEFAULT uuid_generate_v4(),
+    parent_1_name varchar(200),
+    parent_2_name varchar(200),
+    deceased_parent boolean,
+    assent_obtainer varchar(200),
+    consent_id uuid NOT NULL,
+    source_id uuid NOT NULL,
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/db/patches/0104.sql:18]: It was there, I think during a debug I removed but missed to add back. I have added it back.

File: microsetta_private_api/db/patches/0104.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/db/patches/0104.sql:8]: FK is done. Initially I did use a Patch to put the sample doc in DB instead of doing it through test. I was advised to not put any specific sample or data through patch. We do have option to load the DOC in DB using the INSERT in patch file if that is what you both recommend.

File: microsetta_private_api/model/consent.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+                 account_id, reconsent):
+        self.consent_id = consent_id
+        self.consent_type = consent_type
+        self.locale = locale
+        self.date_time = date_time
+        self.consent_content = consent_content
+        self.account_id = account_id
+        self.reconsent = reconsent
+
+    """To be written later!"""
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/model/consent.py:29]: Done

File: microsetta_private_api/model/consent.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class ConsentSignature:
+    @staticmethod
+    def from_dict(input_dict, source_id, signature_id):
+        date_time = datetime.now()
+
+        parent_name_1 = ""
+        if "parent_1_name" in input_dict:
+            parent_name_1 = input_dict["parent_1_name"]
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/model/consent.py:49]: Corrected

File: microsetta_private_api/model/consent.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        self.date_time = date_time
+        self.parent_1_name = parent_1_name
+        self.parent_2_name = parent_2_name
+        self.deceased_parent = deceased_parent
+        self.assent_obtainer = assent_obtainer
+
+    def to_api(self):
+        return {'signature_id': self.signature_id,
+                'consent_id': self.consent_id
+                }
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/model/consent.py:89]: Changed the code to add all properties for the model.

File: microsetta_private_api/repo/account_repo.py [hunk 19]
```diff
@@ -124,11 +124,12 @@ def find_linked_account(self, auth_iss, auth_sub):
                 return AccountRepo._row_to_account(r)
 
     def get_account(self, account_id):
+
         with self._transaction.dict_cursor() as cur:
             cur.execute("SELECT " + AccountRepo.read_cols + " FROM "
                         "account "
                         "WHERE "
-                        "account.id = %s", (account_id,))
+                        "account.id = %s ", (account_id,))
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/repo/account_repo.py:132]: corrected

File: microsetta_private_api/repo/consent_repo.py [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+                          "reconsent_required"
+
+    signature_write_cols = "signature_id, consent_id, " \
+                           "source_id, date_time, parent_1_name, " \
+                           "parent_2_name, deceased_parent, assent_obtainer"
+
+    def create_doc(self, consent):
+        with self._transaction.dict_cursor() as cur:
+            cur.execute("INSERT INTO ag.consent_documents (" +
+                        ConsentRepo.doc_write_cols + ") "
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:78]: corrected

File: microsetta_private_api/repo/consent_repo.py [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+
+        with self._transaction.dict_cursor() as cur:
+            cur.execute("SELECT " + ConsentRepo.doc_read_cols + " FROM "
+                        "consent_documents ORDER BY consent_type DESC, "
+                        "date_time DESC")
+
+            r = cur.fetchall()
+
+            for index in range(0, len(r)):
+                consent_docs.append(_row_to_consent_document(r[index]))
```
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:95]: modified

File: microsetta_private_api/db/patches/0104.sql (hunk 1 above)
Comment (raginirai553) [reply microsetta_private_api/db/patches/0104.sql:17]: Yes, same source will not sign same document twice (consent_id) but all sources of same type, say adult, will have same consent_id( sign same document) so did not add in Consent_audit. However, I added Unique Constraint in the Consent_Document. This way there should never be a situation where a source is flagged for reconsent for same unique id by error.

File: microsetta_private_api/repo/consent_repo.py (hunk 2 above)
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:116]: corrected.

File: microsetta_private_api/repo/consent_repo.py (hunk 3 above)
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:119]: I am trying to find the consent documents signed by a source using a select query with join on consent audit and consent documents table. Preferred to use LIKE to further filter a particular consent document (DATA OR BIOSPECMEN). Avoided the exact test because the age range is not known at the time of performing this query. Though a complex solution would be to extract the age range from source table and programmatically compile exact document type keyword.

File: microsetta_private_api/db/patches/0104.sql (hunk 4 above)
Comment (raginirai553) [reply microsetta_private_api/db/patches/0104.sql:7]: Done

File: microsetta_private_api/repo/consent_repo.py (hunk 5 above)
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:125]: Done

Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:125]: done

File: microsetta_private_api/repo/consent_repo.py (hunk 8 above)
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:144]: Done

File: microsetta_private_api/repo/consent_repo.py (hunk 9 above)
Comment (raginirai553) [reply microsetta_private_api/repo/consent_repo.py:161]: Done

File: microsetta_private_api/db/patches/0104.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/db/patches/0104.sql:8]: ++ When we add the document add function through the admin side things will change anyhow we can apply these recommendation there.

File: microsetta_private_api/db/patches/0104.sql (hunk 15 above)
[the remark this answers is not in the document]
Comment (raginirai553) [reply microsetta_private_api/db/patches/0104.sql:18]: @cassidysymons @wasade disregard the previous comment.I realize the reason why the Foreign Key had to be removed. During the testing I have found that "Deletion of Source" fails if the Foreign Key relation exist.In order to be able to delete source we will need to remove the "Consent_audit" records for the source. Please advise if that is what we want or if there is any other way. All tests pass when the foreign Key relation is removed.
